modules/standard/Socket.chpl

```diff
@@ -0,0 +1,1282 @@
+  :arg timeout: standard ipv6 address.
+  :type timeval: :type:`~Sys.timeval`
+  :return: connected socket.
+  :rtype: `tcpConn`
+  :throws SystemError: On failure to bind or listen on `address`
+*/
+proc listen(in address: ipAddr, reuseAddr: bool = true, backlog: uint(16) = BACKLOG_DEFAULT): tcpListener throws {
+  var family = address.family;
+  var socketFd = socket(family, SOCK_STREAM);
+  bind(socketFd, address, reuseAddr);
+  var err_out = sys_listen(socketFd, backlog:c_int);
+  if err_out != 0 {
+    throw SystemError.fromSyserr(err_out, "Failed to listen on socket");
+  }
+  const listener = new tcpListener(socketFd);
+  return listener;
+}
+
+/*
+  Convenience procedure which creates a :type:`tcpConn` connected to
+  `address`.`timeout` determines how much time to wait for
+  connection to be established. The default value for `timeout` is
+  indefinite.
+
+  .. code-block:: Chapel
+
+    const address = ipAddr.create("127.0.0.1", 8000, IPFamily.IPv4);
+    const timeout = new timeval(4,0);
+    const connectedClient = connect(address, timeout);
+
+  :arg address: address to connect to
+  :type address: :type:`ipAddr`
+  :arg timeout: time to wait for connection establishment.
+  :type timeval: :type:`~Sys.timeval`
+  :return: connected socket.
+  :rtype: `tcpConn`
+  :throws SystemError: Upon failure to connect.
+*/
+proc connect(in address: ipAddr, in timeout = new timeval(-1,0)): tcpConn throws {
+  var family = address.family;
+  var socketFd = socket(family, SOCK_STREAM);
+  var err_out = sys_connect(socketFd, address._addressStorage);
+  if err_out != 0 && err_out != EINPROGRESS {
+    sys_close(socketFd);
+    throw SystemError.fromSyserr(err_out,"connect() failed");
+  }
+  if err_out == 0 {
+    return openfd(socketFd):tcpConn;
+  }
+  var localSync$: sync int = 0;
+  localSync$.readFE();
+  var writerEvent = event_new(event_loop_base, socketFd, EV_WRITE | EV_TIMEOUT | EV_ET, c_ptrTo(syncRWTCallback), c_ptrTo(localSync$):c_void_ptr);
+  defer {
+    event_del(writerEvent);
+    event_free(writerEvent);
+  }
+  if timeout.tv_sec == -1 {
+    err_out = event_add(writerEvent, nil);
+  }
+  else {
+    err_out = event_add(writerEvent, c_ptrTo(timeout));
+  }
+  if err_out != 0 {
+    throw new Error("connect() failed");
+  }
+  var retval = localSync$.readFE();
+  if retval & EV_TIMEOUT != 0 {
+    throw SystemError.fromSyserr(ETIMEDOUT);
+  }
+  err_out = sys_connect(socketFd, address._addressStorage);
+  if err_out != 0 {
+    sys_close(socketFd);
+    throw SystemError.fromSyserr(err_out,"connect() failed");
+  }
+  return openfd(socketFd):tcpConn;
+}
+
+/*
+  This overload of `connect` not only returns a :type:`tcpConn`
+  but also does DNS resolution for the provided `host`.
+  The `timeout` is tired for all resolved addresses and the first
+  successful one is returned back.
+
+  .. code-block:: Chapel
+
+    const timeout = new timeval(4,0);
+    const connectedClient = connect("google.com", "tcp", IPFamily.IPv4, timeout);
```

Comment:
TCP is not a service name - maybe the example would be better with `"http"` for example?